src/tendency.jl

```diff
@@ -154,26 +154,41 @@ end
 
     FT = eltype(q_tot)
 
-    S_q_rai = FT(0)
-    S_q_sno = FT(0)
-    S_q_tot = FT(0)
-    S_q_liq = FT(0)
-    S_q_ice = FT(0)
-    S_q_vap = FT(0) # added for testing
+    S_q_rai::FT = FT(0)
+    S_q_sno::FT = FT(0)
+    S_q_tot::FT = FT(0)
+    S_q_liq::FT = FT(0)
+    S_q_ice::FT = FT(0)
+    S_q_vap::FT = FT(0) # added for testing
 
-    T_fr = KP.T_freeze(params)
-    c_vl = KP.cv_l(params)
-    c_vm = TD.cv_m(thermo_params, ts)
-    Rm = TD.gas_constant_air(thermo_params, ts)
-    Lf = TD.latent_heat_fusion(thermo_params, ts)
+    T_fr::FT = KP.T_freeze(params)
+    c_vl::FT = KP.cv_l(params)
+    c_vm::FT = TD.cv_m(thermo_params, ts)
+    Rm::FT = TD.gas_constant_air(thermo_params, ts)
+    Lf::FT = TD.latent_heat_fusion(thermo_params, ts)
 
     q = TD.PhasePartition(q_tot, q_liq, q_ice)
-    q_vap = TD.vapor_specific_humidity(thermo_params, ts)
+    q_vap::FT = TD.vapor_specific_humidity(thermo_params, ts)
 
     if Bool(params.precip_sources)
+        tmp::FT = FT(0)
         # autoconversion liquid to rain and ice to snow
-        S_qt_rain =
-            -min(max(0.0, q.liq / dt), CM1.conv_q_liq_to_q_rai(microphys_params, q.liq, smooth_transition = true))
+        if params.rain_formation_scheme == "CliMA_1M"
+            tmp = CM1.conv_q_liq_to_q_rai(microphys_params, q.liq, smooth_transition = true)
+        else
+            if params.rain_formation_scheme == "KK2000"
+                tmp = CM2.conv_q_liq_to_q_rai(microphys_params, CMT.KK2000Type(), q.liq, ρ, params.prescribed_Nd)
+            elseif params.rain_formation_scheme == "B1994"
+                tmp = CM2.conv_q_liq_to_q_rai(microphys_params, CMT.B1994Type(), q.liq, ρ, params.prescribed_Nd, true)
+            elseif params.rain_formation_scheme == "TC1980"
+                tmp = CM2.conv_q_liq_to_q_rai(microphys_params, CMT.TC1980Type(), q.liq, ρ, params.prescribed_Nd, true)
+            elseif params.rain_formation_scheme == "LD2004"
+                tmp = CM2.conv_q_liq_to_q_rai(microphys_params, CMT.LD2004Type(), q.liq, ρ, params.prescribed_Nd, true)
+            else
+                error("Unrecognized rain formation scheme")
+            end
+        end
```

Comment:
This is a bit complicated because the arguments for the different methods don't follow a splatt-able pattern between the 1 and 2-moment microphysics schemes.

If they did, I might try doing something like how we call thermo constructors by:

 - Changing `rain_formation_scheme` from a string to one of `CMT.CliMA_1M()`, `CMT.KK2000Type()`,`CMT.B1994Type()`,`CMT.TC1980Type()`,`CMT.LD2004Type()`
 - Define a method to handle which args are needed per type:

```julia
conv_args(::CliMA_1M, ρ, params) = (true, CMT.CliMA_1M())
conv_args(::KK2000Type, ρ, params) = (ρ, params.prescribed_Nd, CMT.KK2000Type())
conv_args(::B1994Type, ρ, params) = (ρ, params.prescribed_Nd, true, CMT.B1994Type())
conv_args(::TC1980Type, ρ, params) = (ρ, params.prescribed_Nd, true, CMT.TC1980Type())
conv_args(::LD2004Type, ρ, params) = (ρ, params.prescribed_Nd, true, CMT.LD2004Type())
```
and call
```julia
tmp = conv_q_liq_to_q_rai(microphys_params, q.liq, conv_args(rain_formation_scheme, ρ, params)...)
```

I suppose we could have `conv_args` return more args to generalize for all calls

Reply:
@charleskawczynski For different `rain_formation_scheme`, `conv_q_liq_to_q_rai` needs to be called from two different modules. Does `tmp = conv_q_liq_to_q_rai(...)` work then?
How about a function instead of conv_args that returns the right rate function:
```
conv_q_liq_to_q_rai(::CliMA_1M, ρ, q.liq, params) = CM1.conv_q_liq_to_q_rai(microphys_params(params), q.liq, smooth_transition = true)
conv_q_liq_to_q_rai(::KK2000Type, ρ, q.liq, params) = CM2.conv_q_liq_to_q_rai(microphys_params(params), CMT.KK2000Type(), q.liq, ρ, params.prescribed_Nd)
```
and so on?

Reply:
> @charleskawczynski For different rain_formation_scheme, conv_q_liq_to_q_rai needs to be called from two different modules. Does tmp = conv_q_liq_to_q_rai(...) work then?

It does if they're imported:

```julia
module modAB

module modA
export foo
struct A end
foo(::A) = 1
end

module modB
export foo
import ..modA: foo
struct B end
foo(::B) = 2
end

end

import .modAB.modA: A, foo
import .modAB.modB: B, foo
foo(A())
foo(B())
```

Reply:
Things already look better with the types and not strings -> only 3 if/else statements.

But I like the idea of hiding it in `conv_args` Will add it and push one more time